core/node/api_server/src/web3/namespaces/debug.rs

```diff
@@ -99,53 +186,48 @@ impl DebugNamespace {
             .get_traces_for_l2_block(block_number)
             .await
             .map_err(DalError::generalize)?;
-        let call_trace = call_traces
-            .into_iter()
-            .map(|call_trace| {
-                let result = Self::map_call(call_trace, only_top_call);
-                ResultDebugCall { result }
-            })
-            .collect();
-        Ok(call_trace)
-    }
 
-    pub async fn debug_trace_block_flat_impl(
-        &self,
-        block_id: BlockId,
-        options: Option<TracerConfig>,
-    ) -> Result<Vec<DebugCallFlat>, Web3Error> {
-        let call_trace = self.debug_trace_block_impl(block_id, options).await?;
-        let call_trace_flat = flatten_debug_calls(call_trace);
-        Ok(call_trace_flat)
+        let mut calls = vec![];
+        let mut flat_calls = vec![];
+        for (call_trace, hash, index) in call_traces {
+            match Self::map_call(call_trace, index, hash, options) {
+                CallTracerResult::CallTrace(call) => calls.push(ResultDebugCall { result: call }),
+                CallTracerResult::FlattCallTrace(mut call) => flat_calls.append(&mut call),
+            }
+        }
+
+        if calls.is_empty() && !flat_calls.is_empty() {
+            Ok(CallTracerBlockResult::FlatCallTrace(flat_calls))
+        } else {
+            Ok(CallTracerBlockResult::CallTrace(calls))
+        }
     }
 
     pub async fn debug_trace_transaction_impl(
         &self,
         tx_hash: H256,
         options: Option<TracerConfig>,
-    ) -> Result<Option<DebugCall>, Web3Error> {
-        let only_top_call = options
-            .map(|options| options.tracer_config.only_top_call)
-            .unwrap_or(false);
+    ) -> Result<Option<CallTracerResult>, Web3Error> {
         let mut connection = self.state.acquire_connection().await?;
         let call_trace = connection
             .transactions_dal()
             .get_call_trace(tx_hash)
             .await
             .map_err(DalError::generalize)?;
-        Ok(call_trace.map(|call_trace| Self::map_call(call_trace, only_top_call)))
+        Ok(call_trace.map(|call_trace| Self::map_call(call_trace, 0, tx_hash, options)))
```

Comment:
Is it OK to return a potentially incorrect `transactionPosition` here? It may make sense to separate the passed trace address index from a transaction index in general; the other case when these two values differ is in `debug_traceBlock*` methods if traces for *some* transactions in a block are absent for whatever reason (although in all fairness, this missing data will likely make the caller upset in any case).